osf_tests/test_registration_machine_states.py

```diff
@@ -0,0 +1,147 @@
+
+        end_date = timezone.now() + settings.EMBARGO_END_DATE_MIN + datetime.timedelta(days=1)
+        draft_registration.run_accept(draft_registration.creator, 'yo', embargo_end_date=end_date)
+
+        draft_registration.save()
+
+        assert draft_registration.machine_state == RegistrationStates.EMBARGO.value
+        assert draft_registration.registered_node
+
+    def test_run_reject(self, draft_registration):
+        draft_registration.run_submit(draft_registration.creator)
+        draft_registration.save()
+        draft_registration.branched_from.save()
+
+        assert draft_registration.machine_state == RegistrationStates.PENDING.value
+
+        draft_registration.run_reject(draft_registration.creator, comment='Double Doink')
+
+        assert draft_registration.machine_state == RegistrationStates.REJECTED.value
+
+    def test_run_request_withdraw(self, draft_registration):
+        draft_registration.run_submit(draft_registration.creator)
+        draft_registration.save()
+
+        draft_registration.run_accept(draft_registration.creator, comment='Wanna Philly Philly?')
+
+        draft_registration.save()
+        assert draft_registration.machine_state == RegistrationStates.ACCEPTED.value
+
+        draft_registration.run_request_withdraw(draft_registration.creator, 'Double Doink')
+
+        assert draft_registration.machine_state == RegistrationStates.PENDING_WITHDRAW.value
+        assert draft_registration.registered_node.retraction.state == Retraction.UNAPPROVED
+
+    def test_run_withdraw_registration(self, draft_registration):
+        draft_registration.run_submit(draft_registration.creator)
+        draft_registration.run_accept(draft_registration.creator, comment='Wanna Philly Philly?')
+        draft_registration.save()
+
+        draft_registration.run_request_withdraw(draft_registration.creator, 'Double Doink')
+
+        assert draft_registration.machine_state == RegistrationStates.PENDING_WITHDRAW.value
+
+        draft_registration.run_withdraw_registration(draft_registration.creator, 'Double Doink')
+
+        assert draft_registration.machine_state == RegistrationStates.WITHDRAWN.value
+        assert draft_registration.registered_node.retraction.state == Retraction.APPROVED
+
+    def test_run_request_embargo_termination(self, draft_registration):
+        draft_registration.run_submit(draft_registration.creator)
+        end_date = timezone.now() + settings.EMBARGO_END_DATE_MIN + datetime.timedelta(days=1)
+        draft_registration.run_accept(draft_registration.creator, 'yo', embargo_end_date=end_date)
+
+        approval_state = draft_registration.registered_node.embargo.approval_state
+        approval_token = approval_state[draft_registration.registered_node.creator._id]['approval_token']
+
+        # This is does not change the state.
+        draft_registration.registered_node.embargo.approve(
+            draft_registration.registered_node.creator,
+            approval_token
+        )
```

Comment:
Is this is a prerequisite for the state change to succeed? If so, it may be worth writing a test without it to ensure that it fails

Reply:
> Is this is a prerequisite for the state change to succeed?

yep embargoes must be approved to be valid, so the state can change.